### ISSUE TITLE
False positive E1136 – unsubscriptable-object

### ISSUE THREAD

**Issue by paulchen5** [pylint-dev/pylint#10455]
### Bug description

This is fine:
```python
end_of_good_plan = good_case_plan_structures.get("<key>", None)
end_of_bad_plan = bad_case_plan_structures.get("<key>", None)

if end_of_good_plan is not None and end_of_bad_plan is not None:
    end_of_good_plan[0]
```

This throws an error:
```python
end_of_good_plan = (
    good_case_plan_structures["<key>"]
    if "<key>" in good_case_plan_structures
    else None
)
end_of_bad_plan = (
    bad_case_plan_structures["<key>"]
    if "<key>" in bad_case_plan_structures
    else None
    )

if end_of_good_plan is not None and end_of_bad_plan is not None:
    end_of_good_plan[0]
```

Useful type hints:
`(variable) good_case_plan_structures: dict[str | None, list[SecretClass]]`
`(variable) bad_case_plan_structures: dict[str | None, list[SecretClass]]`

### Configuration

```ini
default configuration
```

### Command used

```shell
`pylint --disable fixme . `
```

### Pylint output

```python
************* Module src.****.****
src/****/****.py:****:****: E1136: Value 'end_of_good_plan' is unsubscriptable (unsubscriptable-object)
```

### Expected behavior

No error because it's type is being checked

### Pylint version

```shell
pylint 3.3.7
astroid 3.3.10
Python 3.12.0 (main, Dec 23 2024, 00:59:44) [Clang 16.0.0 (clang-1600.0.26.4)]
```

### OS / Environment

macOS 15.5, zsh terminal with oh-my-zsh and p10k theme

### Additional dependencies

```python

```

**Comment by zenlyj** [comment]
Thanks for reporting this, the issue lies in the inference of `end_of_good_plan`, it is being inferred as `None` value in the given example. It is also worth noting that no error is raised when the `if` condition is just a single None check:

```
good_case_plan_structures = {}
bad_case_plan_structures = {}

end_of_good_plan = (
    good_case_plan_structures["<key>"]
    if "<key>" in good_case_plan_structures
    else None
)
end_of_bad_plan = (
    bad_case_plan_structures["<key>"]
    if "<key>" in bad_case_plan_structures
    else None
    )

if end_of_good_plan is not None:
    end_of_good_plan[0]  # no error

if end_of_bad_plan is not None:
    end_of_bad_plan[0]  # no error

if end_of_good_plan is not None and end_of_bad_plan is not None:
    end_of_good_plan[0]  # error, false positive
    end_of_bad_plan[0]  # error, false positive
```

**Comment by Pierre-Sassoulas** [comment]
Astorid 4.0 got some love regarding None comprehension and fixed this.
